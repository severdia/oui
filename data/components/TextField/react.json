{"description":"Generates an `input` element (optionally wrapped in a label) and accepts\nmost of the common input types.","methods":[],"props":{"defaultValue":{"type":{"name":"string"},"required":false,"description":"The default value of the input used on initial render"},"displayError":{"type":{"name":"bool"},"required":false,"description":"Toggle error state styles","defaultValue":{"value":"false","computed":false}},"isDisabled":{"type":{"name":"bool"},"required":false,"description":"Prevents input from being modified and appears disabled","defaultValue":{"value":"false","computed":false}},"isFilter":{"type":{"name":"bool"},"required":false,"description":"Includes error if true","defaultValue":{"value":"false","computed":false}},"isOptional":{"type":{"name":"bool"},"required":false,"description":"Adds an optional label if there is a label provided\n @param {Object} props Object of props\n @returns {Error} Error or null"},"isReadOnly":{"type":{"name":"bool"},"required":false,"description":"Prevents input from being modified but doesn't appear disabled"},"isRequired":{"type":{"name":"bool"},"required":false,"description":"Prevents input from being submitted without value"},"label":{"type":{"name":"string"},"required":false,"description":"Text that describes the input"},"max":{"type":{"name":"number"},"required":false,"description":"Max value for the `input`. Should be used only when `type` is `number`."},"min":{"type":{"name":"number"},"required":false,"description":"Min value for the `input`. Should be used only when `type` is `number`."},"note":{"type":{"name":"string"},"required":false,"description":"Append note near form input.","defaultValue":{"value":"false","computed":false}},"onBlur":{"type":{"name":"func"},"required":false,"description":"Function that fires when the input loses focus. It fires regardless of\nwhether the value has changed."},"onChange":{"type":{"name":"func"},"required":false,"description":"Function that fires when the input loses focus after the value changes"},"onFocus":{"type":{"name":"func"},"required":false,"description":"Function that fires when the input gains focus"},"onInput":{"type":{"name":"func"},"required":false,"description":"Function that fires on keypress"},"onKeyDown":{"type":{"name":"func"},"required":false,"description":"Function that fires when a key is pressed down"},"placeholder":{"type":{"name":"string"},"required":false,"description":"Input placeholder text"},"testSection":{"type":{"name":"string"},"required":false,"description":"Hook for automated JavaScript tests"},"type":{"type":{"name":"enum","value":[{"value":"'text'","computed":false},{"value":"'password'","computed":false},{"value":"'date'","computed":false},{"value":"'number'","computed":false},{"value":"'email'","computed":false},{"value":"'url'","computed":false},{"value":"'search'","computed":false},{"value":"'tel'","computed":false}]},"required":true,"description":"Supported input types"},"value":{"type":{"name":"string"},"required":false,"description":"Text within the input"}},"private":false}