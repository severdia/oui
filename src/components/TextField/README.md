# TextField Component

- Add storybook link
- Add deisgn guidelines