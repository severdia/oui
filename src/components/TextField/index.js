/**
 * DEPERCATION NOTICE: TextField is no longer being maintained,
 * use Input component instead.
 */

import React from 'react';
import Label from '../Label';
import classnames from 'classnames';
import PropTypes from 'prop-types';

/**
 * Generates an `input` element (optionally wrapped in a label) and accepts
 * most of the common input types.
 * @param {Object} props - Properties passed to component
 * @returns {ReactElement}
 */

const TextField = ({
  note = false,
  defaultValue,
  displayError = false,
  isDisabled = false,
  isFilter = false,
  isOptional,
  isReadOnly,
  onBlur,
  isRequired,
  label,
  min,
  max,
  onChange,
  onInput,
  onKeyDown,
  onFocus,
  placeholder,
  testSection,
  type,
  value,
}) => {

  let wrapperClasses = classnames(
    {'oui-form-bad-news': displayError}
  );

  const inputClassNames = classnames(
    'oui-text-input',
    {
      'oui-text-input--search': isFilter,
      'text--right': (type === 'number' || type === 'percent'),
    }
  );

  let inputType = type;
  if (type === 'percent') {
    inputType = 'number';
    // min must be less than max, and both must be within 0 - 100.
    // Otherwise, ignore and set to default
    const oldMin = min;
    min = (!min || min < 0 || min > 100 || min > max) ? 0 : min;
    max = (!max || max < 0 || max > 100 || max < oldMin) ? 100 : max;
  }

  return (
    /* eslint-disable react/jsx-no-bind */
    <div
      data-oui-component={ true }
      className={ wrapperClasses ? 'oui-form-bad-news' : null }>
      { label &&
        <Label
          displayError={ displayError }
          isRequired={ isRequired }
          isOptional={ isOptional }>
          { label }
        </Label>
      }
      <span className="flex flex-align--center">
        <input
          className={ inputClassNames }
          type={ inputType }
          value={ value }
          defaultValue={ defaultValue }
          placeholder={ placeholder }
          required={ isRequired }
          readOnly={ isReadOnly }
          disabled={ isDisabled }
          onInput={ onInput }
          onChange={ onChange }
          onBlur={ onBlur }
          onKeyDown={ onKeyDown }
          onFocus={ onFocus }
          min={ min }
          max={ max }
          data-test-section={ testSection }
        />
        { type === 'percent' && <span className="push-half--left push--right">%</span> }
      </span>
      { note && <div className="oui-form-note">{ note }</div> }
    </div>
  );
};

TextField.propTypes = {
  /** The default value of the input used on initial render */
  defaultValue: PropTypes.string,
  /** Toggle error state styles  */
  displayError: PropTypes.bool,
  /** Prevents input from being modified and appears disabled */
  isDisabled: PropTypes.bool,
  /** Includes error if true */
  isFilter: PropTypes.bool,
  /** Adds an optional label if there is a label provided
   *  @param {Object} props Object of props
   *  @returns {Error} Error or null
   */
  isOptional: PropTypes.bool,
  /** Prevents input from being modified but doesn't appear disabled */
  isReadOnly: PropTypes.bool,
  /** Prevents input from being submitted without value */
  isRequired: PropTypes.bool,
  /** Text that describes the input */
  label: PropTypes.string,
  /**
   * Max value for the `input`. Should be used only when `type` is `number`.
   */
  max: PropTypes.number,
  /**
   * Min value for the `input`. Should be used only when `type` is `number`.
   */
  min: PropTypes.number,
  /** Append note near form input. */
  note: PropTypes.string,
  /**
   * Function that fires when the input loses focus. It fires regardless of
   * whether the value has changed.
  */
  onBlur: PropTypes.func,
  /** Function that fires when the input loses focus after the value changes */
  onChange: PropTypes.func,
  /** Function that fires when the input gains focus */
  onFocus: PropTypes.func,
  /** Function that fires on keypress */
  onInput: PropTypes.func,
  /** Function that fires when a key is pressed down */
  onKeyDown: PropTypes.func,
  /** Input placeholder text */
  placeholder: PropTypes.string,
  /** Hook for automated JavaScript tests */
  testSection: PropTypes.string,
  /** Supported input types */
  type: PropTypes.oneOf([
    'text',
    'password',
    'date',
    'number',
    'percent',
    'email',
    'url',
    'search',
    'tel',
  ]).isRequired,
  /** Text within the input */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default TextField;
