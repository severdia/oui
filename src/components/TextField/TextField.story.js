import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, boolean, number, select } from '@storybook/addon-knobs';

import TextField from './index.js';

const stories = storiesOf('TextField (deprecated)', module);
stories
  .addDecorator(withKnobs)
  .addDecorator(story => (
    <div id="root-preview">
      <div className="attention attention--bad-news oui-text--center push-double--bottom">
        TextField component is deprecated. Please use Input instead for the same features.
      </div>
      {story()}
    </div>
  ));

stories
  .add('TextField with knobs', (() => {
    return (
      <TextField
        defaultValue={ text('defaultValue', 'some default value') }
        displayError={ boolean('displayError', false) }
        isFilter={ boolean('isFilter', false) }
        isOptional={ boolean('isOptional', false) }
        isRequired={ boolean('isRequired', false) }
        label={ text('Label', 'Use Input Component instead') }
        max={ number('max', 50) }
        min={ number('min', 10) }
        note={ text('note', '') }
        onChange={ action('on change') }
        onBlur={ action('on blur') }
        onKeyDown={ action('on key press') }
        placeholder={ text('placeholder', 'just a placeholder') }
        type={ select('type', ['text', 'password', 'number', 'date', 'percent'], 'text') }
      />);
  })).add('TextField', () => {
    return (
      <TextField
        label="Use Input Component instead"
        note="A short description or note about this field."
        placeholder="Just a placeholder"
        type="text"
      />);
  }).add('Percent', () => {
    return (
      <TextField
        label="Total Traffic Allocated:"
        placeholder="100"
        defaultValue={ text('defaultValue', '0') }
        type="percent"
        max={ number('max', 100) }
        min={ number('min', 0) }
      />);
  }).add('Error state', () => {
    return (
      <TextField
        label="Use Input Component instead"
        displayError={ true }
        note="Form error note."
        placeholder="Just a placeholder"
        type="text"
      />);
  });
