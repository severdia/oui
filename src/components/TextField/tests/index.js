import React from 'react';
import TextField from '../index';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

describe('components/TextField', () => {
  it('should render a simple input with type text', () => {
    const output = shallow(
      <TextField
        type='text'
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should render with a simple label', () => {
    const output = shallow(
      <TextField
        label='simple label'
        type='text'
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should render with a simple label and a note', () => {
    const output = shallow(
      <TextField
        label='Simple label'
        note='simple note'
        type='text'
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should render error style', () => {
    const output = shallow(
      <TextField
        displayError={ true }
        label='Simple label'
        note='simple note'
        type='text'
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should render with a placeholder', () => {
    const output = shallow(
      <TextField
        placeholder='Simple placeholder'
        type='text'
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should render search icon', () => {
    const output = shallow(
      <TextField
        isFilter={ true }
        type='text'
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should render max and min limits', () => {
    const output = shallow(
      <TextField
        min={ 0 }
        max={ 100 }
        type='number'
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should call onChange fn() prop when a change happen on input', () => {
    const blurEvent = jest.fn();
    const textField = mount(
      <TextField
        type='text'
        onBlur={ blurEvent }
      />
    );
    const input = textField.find('.oui-text-input');
    input.simulate('blur');
    expect(blurEvent).toHaveBeenCalled();
  });

  it('should call onBlur fn() prop when clicking out of input', () => {
    const changeEvent = jest.fn();
    const textField = mount(
      <TextField
        onChange={ changeEvent }
        type='text'
      />
    );
    const input = textField.find('.oui-text-input');
    input.simulate('change');
    expect(changeEvent).toHaveBeenCalled();
  });

  it('should call onKeyDown fn() prop when pressing keys on input', () => {
    const keydownEvent = jest.fn();
    const textField = mount(
      <TextField
        onKeyDown={ keydownEvent }
        type='text'
      />
    );
    const input = textField.find('.oui-text-input');
    input.simulate('keyDown');
    expect(keydownEvent).toHaveBeenCalled();
  });

  it('should call onFocus fn() prop when clicking on input', () => {
    const focusEvent = jest.fn();
    const textField = mount(
      <TextField
        onFocus={ focusEvent }
        type='text'
      />
    );
    const input = textField.find('.oui-text-input');
    input.simulate('focus');
    expect(focusEvent).toHaveBeenCalled();
  });

  it('should render HTML type=number for when prop.type is percent', () => {
    const output = shallow(
      <TextField
        type='percent'
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  describe('custom min max percentages', function() {

    it('should allow minimum and maximum percentages when type is percent', () => {
      const output = shallow(
        <TextField
          type='percent'
          min={ 70 }
          max={ 85 }
        />
      );
      expect(output.find('input').props().min).toBe(70);
      expect(output.find('input').props().max).toBe(85);
      expect(shallowToJson(output)).toMatchSnapshot();
    });

    describe('invalid custom minimum', function() {

      it('should revert to the default if it is below 0', () => {
        const output = shallow(
          <TextField
            type='percent'
            min={ -10 }
          />
        );
        expect(output.find('input').props().min).toBe(0);
        expect(output.find('input').props().max).toBe(100);
        expect(shallowToJson(output)).toMatchSnapshot();
      });

      it('should revert to the default if it is above 100', () => {
        const output = shallow(
          <TextField
            type='percent'
            min={ 101 }
          />
        );
        expect(output.find('input').props().min).toBe(0);
        expect(output.find('input').props().max).toBe(100);
        expect(shallowToJson(output)).toMatchSnapshot();
      });

      it('should revert to the default if it is above the maximum', () => {
        const output = shallow(
          <TextField
            type='percent'
            min={ 80 }
            max={ 79 }
          />
        );
        expect(output.find('input').props().min).toBe(0);
        expect(output.find('input').props().max).toBe(100);
        expect(shallowToJson(output)).toMatchSnapshot();
      });
    });

    describe('invalid custom maximum', function() {

      it('should revert to the default if it is below 0', () => {
        const output = shallow(
          <TextField
            type='percent'
            max={ -10 }
          />
        );
        expect(output.find('input').props().min).toBe(0);
        expect(output.find('input').props().max).toBe(100);
        expect(shallowToJson(output)).toMatchSnapshot();
      });

      it('should revert to the default if it is above 100', () => {
        const output = shallow(
          <TextField
            type='percent'
            max={ 101 }
          />
        );
        expect(output.find('input').props().min).toBe(0);
        expect(output.find('input').props().max).toBe(100);
        expect(shallowToJson(output)).toMatchSnapshot();
      });

      it('should revert to the default if it is below the minimum', () => {
        const output = shallow(
          <TextField
            type='percent'
            min={ 80 }
            max={ 79 }
          />
        );
        expect(output.find('input').props().min).toBe(0);
        expect(output.find('input').props().max).toBe(100);
        expect(shallowToJson(output)).toMatchSnapshot();
      });
    });
  });
});
