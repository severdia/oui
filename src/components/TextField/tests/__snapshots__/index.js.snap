// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TextField custom min max percentages invalid custom maximum should revert to the default if it is above 100 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input text--right"
      disabled={false}
      max={100}
      min={0}
      type="number"
    />
    <span
      className="push-half--left push--right"
    >
      %
    </span>
  </span>
</div>
`;

exports[`components/TextField custom min max percentages invalid custom maximum should revert to the default if it is below 0 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input text--right"
      disabled={false}
      max={100}
      min={0}
      type="number"
    />
    <span
      className="push-half--left push--right"
    >
      %
    </span>
  </span>
</div>
`;

exports[`components/TextField custom min max percentages invalid custom maximum should revert to the default if it is below the minimum 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input text--right"
      disabled={false}
      max={100}
      min={0}
      type="number"
    />
    <span
      className="push-half--left push--right"
    >
      %
    </span>
  </span>
</div>
`;

exports[`components/TextField custom min max percentages invalid custom minimum should revert to the default if it is above 100 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input text--right"
      disabled={false}
      max={100}
      min={0}
      type="number"
    />
    <span
      className="push-half--left push--right"
    >
      %
    </span>
  </span>
</div>
`;

exports[`components/TextField custom min max percentages invalid custom minimum should revert to the default if it is above the maximum 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input text--right"
      disabled={false}
      max={100}
      min={0}
      type="number"
    />
    <span
      className="push-half--left push--right"
    >
      %
    </span>
  </span>
</div>
`;

exports[`components/TextField custom min max percentages invalid custom minimum should revert to the default if it is below 0 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input text--right"
      disabled={false}
      max={100}
      min={0}
      type="number"
    />
    <span
      className="push-half--left push--right"
    >
      %
    </span>
  </span>
</div>
`;

exports[`components/TextField custom min max percentages should allow minimum and maximum percentages when type is percent 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input text--right"
      disabled={false}
      max={85}
      min={70}
      type="number"
    />
    <span
      className="push-half--left push--right"
    >
      %
    </span>
  </span>
</div>
`;

exports[`components/TextField should render HTML type=number for when prop.type is percent 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input text--right"
      disabled={false}
      max={100}
      min={0}
      type="number"
    />
    <span
      className="push-half--left push--right"
    >
      %
    </span>
  </span>
</div>
`;

exports[`components/TextField should render a simple input with type text 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input"
      disabled={false}
      type="text"
    />
  </span>
</div>
`;

exports[`components/TextField should render error style 1`] = `
<div
  className="oui-form-bad-news"
  data-oui-component={true}
>
  <Label
    displayError={true}
  >
    Simple label
  </Label>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input"
      disabled={false}
      type="text"
    />
  </span>
  <div
    className="oui-form-note"
  >
    simple note
  </div>
</div>
`;

exports[`components/TextField should render max and min limits 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input text--right"
      disabled={false}
      max={100}
      min={0}
      type="number"
    />
  </span>
</div>
`;

exports[`components/TextField should render search icon 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input oui-text-input--search"
      disabled={false}
      type="text"
    />
  </span>
</div>
`;

exports[`components/TextField should render with a placeholder 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input"
      disabled={false}
      placeholder="Simple placeholder"
      type="text"
    />
  </span>
</div>
`;

exports[`components/TextField should render with a simple label 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <Label
    displayError={false}
  >
    simple label
  </Label>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input"
      disabled={false}
      type="text"
    />
  </span>
</div>
`;

exports[`components/TextField should render with a simple label and a note 1`] = `
<div
  className={null}
  data-oui-component={true}
>
  <Label
    displayError={false}
  >
    Simple label
  </Label>
  <span
    className="flex flex-align--center"
  >
    <input
      className="oui-text-input"
      disabled={false}
      type="text"
    />
  </span>
  <div
    className="oui-form-note"
  >
    simple note
  </div>
</div>
`;
